### ISSUE TITLE
[HOLD for payment 2024-01-26] [$500] Emoji - When resizing the window, the open emoji picker moves to the LHN side

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#31155]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.97-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Open https://staging.new.expensify.com/
2. Log in with any account
3. Open any report
4. In Compose Box, open the emoji picker
5. Reduce the width of the open browser window

## Expected Result:
Emoji picker should be fixed above the Compose Box when resizing the window and should remain open when going into mobile mode

## Actual Result:
When resizing the window, the open emoji picker moves to the LHN side

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/746dacd2-39bb-4e46-9d00-ef92c43d2636

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b9e60337c2ef461a</li>
        <li>Upwork Job ID: 1722692276888514560</li>
        <li>Last Price Increase: 2023-11-30</li>
    </ul>
</details>

**Comment by jeet-dhandha** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- Emoji picker sticking to left side of screen when we resize the window horizontally.

### What is the root cause of that problem?
- In below code snippet the value comes as `{horizontal: 0, vertical: 0}` when we resize the window horizontally.

https://github.com/Expensify/App/blob/e2c1c031d388b6e7277249d87d3e13197f2cbf3a/src/components/EmojiPicker/EmojiPicker.js#L132

- This happens because when we copy reference of anchor element and then measure it's position using `measureInWindow` the response of callback comes as (x, y, width, height) => (0, 0, 0, 0) when we resize the window horizontally.

- And if we check the original anchor element's position using `measure` it's position is correct.

### What changes do you think we should make in order to solve the problem?
- First we create a `useDimensionChange` hook which will take a callback function and will call it whenever the dimension of the element changes.
- then we move the calculate logic to a function `onAnchorDimensionChange` which takes the original ref and pass it to `useImperativeHandle`.

https://github.com/Expensify/App/blob/e2c1c031d388b6e7277249d87d3e13197f2cbf3a/src/components/EmojiPicker/EmojiPicker.js#L124-L135

- then we create a `onAnchorDimensionChange` function in `EmojiPickerAction.ts` and call it in `useDimensionChange` at each place where we use the `showEmojiPicker` function.

Below example is for `EmojiPickerButton.js` but we have to do the same for other places where we use `showEmojiPicker` function.
```js
// `emojiPopoverAnchor.current` is the original ref of the anchor element from `EmojiPickerButton.js`
useDimensionChange(() => EmojiPickerAction.onAnchorDimensionChange(emojiPopoverAnchor.current));
```

### What alternative solutions did you explore? (Optional)
- N/A

**Comment by dragnoir** [comment]
Contributor details
Your Expensify account email: [mosaixel.org@gmail.com](mailto:mosaixel.org@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/rachidlajmi

**Comment by dragnoir** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When resizing the window on the web browser, the open emoji picker moves to the LHN side.

### What is the root cause of that problem?

The issue is exclusively observed in web browsers when users interact using a mouse. I conducted tests on a touchscreen device and encountered no problems. Further testing in Google Chrome's DevTools using the 'toggle device' option and responsive layout modes also did not reproduce the issue.

The specific problem arises when a user clicks the emoji button. This action triggers the addition of a `background` style to the anchor element in the DOM. Our `calculateAnchorPosition` function retrieves the `emojiPopoverAnchor.current` element, which includes this background style. However, by the time `measureInWindow` initiates its search for this element, the DOM has already updated to a version where the anchor no longer has the background style. This discrepancy is likely causing the issue we're observing.

### What changes do you think we should make in order to solve the problem?
To address the issue, we can implement a detection mechanism to ascertain if the user is interacting via a mouse. In scenarios where the window is being resized, we can strategically hide the Emoji Picker. This approach maintains the existing user experience on touchscreen devices, while effectively resolving the issue on desktops and laptops. By hiding the Emoji Picker during window resizing and only displaying it again upon user click, we ensure a seamless experience across different devices.

```js
// state for user using a mouse
const [isUsingMouse, setIsUsingMouse] =  useState(false);

useEffect(() => {
	const  handleMouseMove  = () => {
		setIsUsingMouse(true);
		// Remove the event listener after detecting mouse movement
		window.removeEventListener('mousemove', handleMouseMove);
	};

	if (!isUsingMouse) {
		window.addEventListener('mousemove', handleMouseMove);
	}

	const  emojiPopoverDimensionListener  =  Dimensions.addEventListener('change', () => {
		if (isUsingMouse) {
			hideEmojiPicker();
		}
		// rest of the code ..
		return () => {
			emojiPopoverDimensionListener.remove();
			window.removeEventListener('mousemove', handleMouseMove);
	};
}, [isEmojiPickerVisible, isSmallScreenWidth, emojiPopoverAnchorOrigin]);
```

### What alternative solutions did you explore? (Optional)

-   N/A

**Comment by dragnoir** [comment]
Contributor details
Your Expensify account email: mosaixel.org@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/rachidlajmi

**Comment by dragnoir** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When resizing the window on the web browser, the open emoji picker moves to the LHN side.

### What is the root cause of that problem?
I was reviewing my last proposal and with more testing, it seems that I am wrong, it's not about the style on the DOM. I checked more about the element position and the best practice to it and I found that EmojiPickerButton is not structured the way it's recommanded for `measureInWindow`.
There's some rendering issues that prevent the function from getting the right anchor details.

### What changes do you think we should make in order to solve the problem?
The return of EmojiPickerButton.js should be embedded inside a view where the refs should be.

```js
import {View} from  'react-native';

return (
	<View
		ref={emojiPopoverAnchor}
	>
		<Tooltip  text={props.translate('reportActionCompose.emoji')}>
			<PressableWithoutFeedback
	// rest of the code
			</PressableWithoutFeedback>
		</Tooltip>
	</View>
);
```
so moving `ref={emojiPopoverAnchor}` from PressableWithoutFeedback to the new View will solve the issue.

### What alternative solutions did you explore? (Optional)

-   N/A

**Comment by lschurr** [comment]
Any update @aimane-chnaif?

**Comment by dragnoir** [comment]
@aimane-chnaif Have you had a chance to review my second proposal pls!
It also solves another issue I have seen on the video from "lanitochka17" when the screen is small (mobile view) and we make it wider, the popup does not close.

I may disagree with "bernhardoj", for the other component that uses the same utility to calculate the position, there's no issue, but they get the ref from a View as I mentioned in my second proposal.

![image](https://github.com/Expensify/App/assets/12425932/86371583-1bd7-463e-9e96-55c34bf68f9c)

**Comment by dragnoir** [comment]
@aimane-chnaif it's late here, I will send a reply to your question tomorrow.

**Comment by dragnoir** [comment]
@aimane-chnaif I reviewed, this morning, the additional components you referenced and noted that they do not recalculate the anchor's new position when there is a change in screen size.

There are no issues with the "`PopoverAnchorTooltip`" code or any other related functions. Furthermore, if you test the same behavior on a tablet or a touchscreen device, you will not encounter the problem; the new position is accurately recalculated.

The issue arises when there is a change in screen size on web. The "`EmojiPicker`" components require the updated position of the anchor (the button that was clicked). We use "measureInWindow" from React Native to determine the new coordinates (x, y) of the anchor.

I discovered similar cases on "Stack Overflow" where individuals faced incorrect positioning issues upon screen resizing. These were resolved using my second suggestion.

The problem with "measureInWindow" is that it requires the reference (ref) of the View as a prop, and this reference must be inherited from a View. If you examine "EmojiPickerButton.js", you'll see that the anchor element's ref is set within "`PressableWithoutFeedback`", which does not align with how "`measureInWindow`" functions.
https://github.com/Expensify/App/blob/72e73ae6d922b915198e0a5bcff9d71de5a1001b/src/libs/calculateAnchorPosition.ts#L20

Regarding the issues raised by "bernhardoj", my proposed solution addresses them as well. In "`AddReactionBubble.js`", using a View for the reference that will be passed to the function calculating the new position will resolve these issues.
https://github.com/Expensify/App/blob/72e73ae6d922b915198e0a5bcff9d71de5a1001b/src/components/Reactions/AddReactionBubble.js#L88-L89

```js
<View  ref={ref}> // ref placed here
	<Tooltip  text={props.translate('emojiReactions.addReactionTooltip')}>
		<PressableWithFeedback
			// ref removed from here 
			style={({hovered, pressed}) => ...
```
In summary, I see that there is nothing incorrect with the existing code. The challenge lies in the fact that "measureInWindow" requires a View reference, as detailed in the React Native documentation.

**Comment by lschurr** [comment]
@aimane-chnaif can you check the proposals here?

**Comment by dragnoir** [comment]
@aimane-chnaif Ok I will check the solution won't cause any style regressions, I will reply ASAP

**Comment by dragnoir** [comment]
@aimane-chnaif, I've thoroughly reviewed my implementation and confirmed that adding an additional View component does not lead to any style regressions.

@bernhardoj, thank you for your time in reviewing my solution. I've taken the opportunity to evaluate your suggestion and would like to share some observations:
- The method you've employed to obtain the correct reference seems overly complex.
- This approach does not align with the best practices recommended in the React Native documentation.
- Upon integrating your proposal, I encountered functionality issues with other components, such as the EmojiPicker used in comment reactions.
- I'd like to assure you that introducing an additional view in my solution does not result in unintended margins or alterations to existing styles.

**Comment by lschurr** [comment]
@aimane-chnaif can you review?

**Comment by lschurr** [comment]
Bumping in Slack - https://expensify.slack.com/archives/C01GTK53T8Q/p1701964385472649

**Comment by lschurr** [comment]
Can you review @aimane-chnaif?

**Comment by lschurr** [comment]
Bumped in Slack - https://expensify.slack.com/archives/C01GTK53T8Q/p1702537717765929?thread_ts=1701964385.472649&cid=C01GTK53T8Q

**Comment by lschurr** [comment]
Bumped the Slack thread.

**Comment by allroundexperts** [comment]
Just went through the whole conversation. If I was implementing this, I would go via the solution that @bernhardoj has proposed. The ugliness can be supplemented by using a getter ref as suggested by @bernhardoj.

I can see that this is tricky but @tienifr's solution looks too complex to me. @dragnoir's solution is a workaround, I agree with that.

**Comment by neil-marcellini** [comment]
> Let's go with @bernhardoj's [proposal](https://github.com/Expensify/App/issues/31155#issuecomment-1805431645) 🎀 👀 🎀 C+ reviewed

I agree. @tienifr regarding your comment, I like to generally agree with what the C+ thinks. I looked at your proposal and I think it's pretty cool, but it's not the most simple thing. I like to go with the most simple solution that is easiest for most people to understand.

**Comment by lschurr** [comment]
Hi @aimane-chnaif - could you work on the checklist for this one?

**Comment by lschurr** [comment]
Payment summary:

- Bug reporter: Applause (no payment)
- Contributor: @bernhardoj $500
- Contributor+: @aimane-chnaif $500

**Comment by lschurr** [comment]
Could you accept the offers sent?

@bernhardoj: https://www.upwork.com/nx/wm/offer/100280589
@aimane-chnaif: https://www.upwork.com/nx/wm/offer/100280580

**Comment by bernhardoj** [comment]
We still have a small regression [here](https://github.com/Expensify/App/pull/33690). @aimane-chnaif Can you please check?

**Comment by lschurr** [comment]
Waiting to issue payments for the regression. Bump @aimane-chnaif

**Comment by trjExpensify** [comment]
Apparently this is also a regression from this issue: https://github.com/Expensify/App/issues/33873#issuecomment-1876784916

**Comment by lschurr** [comment]
Bumped @aimane-chnaif in Slack as well - https://expensify.slack.com/archives/C01GTK53T8Q/p1704732500083039

**Comment by lschurr** [comment]
Still waiting on the regression.

**Comment by lschurr** [comment]
Hi @aimane-chnaif @bernhardoj @neil-marcellini - Could someone take a look here and let me know how to proceed? Has the regression been fixed?

**Comment by lschurr** [comment]
Any update @aimane-chnaif?

**Comment by neil-marcellini** [comment]
Yep, under review.

**Comment by lschurr** [comment]
@aimane-chnaif - Can you work through the checklist on this one?

**Comment by lschurr** [comment]
It looks like the automation didn't work here. Moving to Daily and will handle payments shortly.

**Comment by lschurr** [comment]
Payment summary:

- Bug reporter: Internal (no payment)
- Contributor: @bernhardoj $250
- Contributor+: @aimane-chnaif $250

**Comment by lschurr** [comment]
Offers sent in Upwork.

**Comment by lschurr** [comment]
Ah, sorry about that! I'll use the previous offers.
